File: slicing-ios/SlicingGame/Page/Modules/BasicModules/VibrateModule.swift
//
//  VibrateModule.swift
//  Basic module: catches events to vibrate the device
//

import UIKit
import JsonInflator
import AudioToolbox

class VibrateModule: PageModule {
    
    // --
    // MARK: Vibrate
    // --
    
    enum VibrateImpact {
        
        case light
        case normal
        case heavy
        
    }
    
    class func vibrate(impact: VibrateImpact) {
        if #available(iOS 10.0, *) {
            let style: UIImpactFeedbackGenerator.FeedbackStyle
            switch impact {
            case .light:
                style = .light
            case .normal:
                style = .medium
            case .heavy:
                style = .heavy
            }
            UIImpactFeedbackGenerator(style: style).impactOccurred()
        } else {
            AudioServicesPlaySystemSound(kSystemSoundID_Vibrate)
        }
    }


    // --
    // MARK: Members
    // --
    
    let handleEventTypes = ["vibrate"]

    
    // --
    // MARK: Inflator integration
    // --
    
    class func inflatable() -> JsonInflatable {
        return InflatorClass()
    }
    
    private class InflatorClass: JsonInflatable {
        
        func create() -> Any {
            return VibrateModule()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            return object is VibrateModule
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return type(of: object) == VibrateModule.self
        }
        
    }


    // --
    // MARK: Initialization
    // --
    
    init() {
        // No implementation
    }


    // --
    // MARK: Lifecycle
    // --

    func didCreate(viewController: PageViewController) {
        // No implementation
    }
    
    func didResume() {
        // No implementation
    }
    
    func didPause() {
        // No implementation
    }
    
    func willDestroy() {
        // No implementation
    }
    

    // --
    // MARK: Event handling
    // --

    func handleEvent(_ event: AppEvent, sender: Any?) -> Bool {
        if event.type == "vibrate" {
            VibrateModule.vibrate(impact: .light)
            return true
        }
        return false
    }

}

File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/page/modules/basicmodules/VibrateModule.kt
package com.crescentflare.slicinggame.page.modules.basicmodules

import android.content.Context
import android.content.Context.VIBRATOR_SERVICE
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.slicinggame.infrastructure.events.AppEvent
import com.crescentflare.slicinggame.page.activities.PageActivity
import com.crescentflare.slicinggame.page.modules.PageModule
import java.lang.ref.WeakReference

/**
 * Basic module: catches events to vibrate the device
 */
class VibrateModule: PageModule {

    // --
    // Statics
    // --

    companion object {

        // --
        // Static: inflatable integration
        // --

        val inflatable: JsonInflatable = object : JsonInflatable {

            override fun create(context: Context): Any {
                return VibrateModule()
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                return obj is VibrateModule
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj::class == VibrateModule::class
            }

        }


        // --
        // Static: vibrate
        // --

        fun vibrate(context: Context, duration: Int) {
            val vibrator = context.getSystemService(VIBRATOR_SERVICE)
            (vibrator as? Vibrator)?.let {
                if (Build.VERSION.SDK_INT >= 26) {
                    it.vibrate(VibrationEffect.createOneShot(duration.toLong(), VibrationEffect.DEFAULT_AMPLITUDE));
                } else {
                    it.vibrate(duration.toLong());
                }
            }
        }

    }


    // --
    // Members
    // --

    override val handleEventTypes = listOf("vibrate")
    private var activityReference : WeakReference<PageActivity>? = null


    // --
    // Lifecycle
    // --

    override fun onCreate(activity: PageActivity) {
        activityReference = WeakReference(activity)
    }

    override fun onResume() {
        // No implementation
    }

    override fun onPause() {
        // No implementation
    }

    override fun onDestroy() {
        // No implementation
    }

    override fun onLowMemory() {
        // No implementation
    }

    override fun onBackPressed(): Boolean {
        return false // No custom handling for the back button
    }


    // --
    // Event handling
    // --

    override fun handleEvent(event: AppEvent, sender: Any?): Boolean {
        if (event.type == "vibrate") {
            activityReference?.get()?.let {
                vibrate(it, 60)
            }
            return true
        }
        return false
    }

}
